scipy/linalg/decomp.py

```diff
@@ -1267,3 +1267,138 @@ def hessenberg(a, calc_q=False, overwrite_a=False, check_finite=True):
+    For example, obtained by `w, v = scipy.linalg.eig(X)` or
+    `w, v = numpy.linalg.eig(X)` in which case `X` can also represent
+    stacked arrays.
+
+    Examples
+    --------
+    >>> import numpy
+    >>> X = numpy.array([[1, 2, 3], [0, 4, 5], [0, -5, 4]])
+    >>> X
+    array([[ 1,  2,  3],
+           [ 0,  4,  5],
+           [ 0, -5,  4]])
+
+    >>> from scipy import linalg
+    >>> w, v = linalg.eig(X)
+    >>> w
+    array([ 1.+0.j,  4.+5.j,  4.-5.j])
+    >>> v
+    array([[ 1.00000+0.j     , -0.01906-0.40016j, -0.01906+0.40016j],
+           [ 0.00000+0.j     ,  0.00000-0.64788j,  0.00000+0.64788j],
+           [ 0.00000+0.j     ,  0.64788+0.j     ,  0.64788-0.j     ]])
+
+    >>> wr, vr = linalg.cdf2rdf(w, v)
+    >>> wr
+    array([[ 1.,  0.,  0.],
+           [ 0.,  4.,  5.],
+           [ 0., -5.,  4.]])
+    >>> vr
+    array([[ 1.     ,  0.40016, -0.01906],
+           [ 0.     ,  0.64788,  0.     ],
+           [ 0.     ,  0.     ,  0.64788]])
+
+    >>> vr @ wr
+    array([[ 1.     ,  1.69593,  1.9246 ],
+           [ 0.     ,  2.59153,  3.23942],
+           [ 0.     , -3.23942,  2.59153]])
+    >>> X @ vr
+    array([[ 1.     ,  1.69593,  1.9246 ],
+           [ 0.     ,  2.59153,  3.23942],
+           [ 0.     , -3.23942,  2.59153]])
+    """
+    w1 = _asarray_validated(w)
+    v1 = _asarray_validated(v)
+    if len(w1.shape) < 1 or len(w1.shape) > 2:
+        raise ValueError('expected 1D array or stacked 1D arrays')
+    if (len(v1.shape) < 2 or len(v1.shape) > 3
+        or v1.shape[-2] != v1.shape[-1]
+        or v1.shape[-1] < 2):
+        raise ValueError('expected square matrix or stacked square matrices')
+    if len(v1.shape) != len(w1.shape) + 1 or v1.shape[-1] != w1.shape[-1]:
+        raise ValueError('expected eigenvalues and associated eigenvectors')
+
+    # number of stacked matrices and number of eigenvalues in each matrix
+    if w.ndim == 1:
+        stacked = False
+        M, n = 1, len(w)
+        w = w[numpy.newaxis, :]
+        v = v[numpy.newaxis, :, :]
+    else:
+        stacked = True
+        M, n = w.shape
+
+    # get indices for each first pair of complex eigenvalues
+    idx_im = numpy.argwhere(numpy.iscomplex(w))
+
+    # check if all complex eigenvalues have conjugate pairs
+    if len(idx_im[::2, 0]) != len(idx_im[1::2, 0]):
+        raise ValueError('expected complex-conjugate pairs of eigenvalues')
+
+    # all eigenvalues to diagonal form
+    w_mat = numpy.zeros((M, n, n))
+    di = range(n)
+    w_mat[:, di, di] = w[:, di].real
+
+    # complex eigenvalues to real block diagonal form
+    mat, k = idx_im[::2, 0], idx_im[::2, 1]
+    w_mat[mat, k, k+1] = w[mat, k].imag
+    w_mat[mat, k+1, k] = w[mat, k+1].imag
+    wr = w_mat.real
+
+    # compute real eigenvectors associated with real block diagonal eigenvalues
+    u = numpy.array([numpy.eye(n, n,
+                    dtype=numpy.complex64)] * M).reshape(M, n, n)
+    u[mat, k, k] = 0.5j
```

Comment:
I think it will be a little more efficient to do what I did in my modified code at https://github.com/scipy/scipy/issues/8544:  precompute the 2x2 array `U`, and assign the block to `u` in one statement: `u[mat, k:k+2, k:k+2] = U`.  Even if the efficiency gain is negligible, I think it makes the code easier to read.

Reply:
I agree that your suggestion would be better; however, this only works if we have a single eigenvector array, for stacked arrays, k is not int, but array, and this indexing does not work (I receive TypeError: only integer scalar arrays can be converted to a scalar index)